### ISSUE TITLE
Inconsistent use of outputs and inputs+outputs in core resource model

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#2650]
As discussed in https://github.com/pulumi/pulumi-terraform/issues/362#issuecomment-482853828, the engine currently passes only outputs to Diff, but passes a merged inputs+outputs to Update.  This can lead to different deltas being observed in each case, ultimately violating presumed invariants that providers should be able to assume - in particular, that they have had a chance to review and define replacements in the call to Diff prior to processing the resulting change in Update.  In that issue, this was able to lead to different changes being seen - such that a Pulumi update lead to a Terraform provider Delete.  This has been fixed, but the core engine inconsistency and potential for problems for providers still exists.

There are several questions/issues here:
1. Is it correct to ever merge inputs+outputs in the engine?  What does this actually mean, and how should provider be expect to reason about it?
2. If it is correct to merge, what are reasonable semantics for the merge?  In particular, the current merge recursively elementwise merges arrays if it finds them in the merge.  I cannot reason about how that could ever make sense (the merged array is not necessary either of the input or output) - but I don't know the answer to (1) so it's possible there's a scenario/requirement. 
3. If it is correct to merge, is it correct to *not* merge for Diff?  It seems like this cannot be correct for any reasonable invariants that could be assumed by the provider - since it has to have the chance to define the replacements before trying to process updates (in particular, it should be able to avoid having any code to handle updates for changes related to properties it indicated requires replacement - which the current rule allows the engine to violate).

**Comment by lukehoban** [comment]
In https://github.com/pulumi/pulumi-eks/issues/69, we have a case where the `input` and `output` array values are in different orders, and then the element-wise merge creates bogus merged array values (hybrids of two of the actual values).

The underlying provider knows how to properly handle this array as a set, but an element-wise merge does not.  This seems yet-another reason to not trust the merge (at least for arrays).

**Comment by joeduffy** [comment]
There is a long history to this one.

TL;DR, the use of `All()` and `Merge()` is almost certainly wrong, and we fixed a similar class of bugs over a year ago, where we had been merging _inputs_, in https://github.com/pulumi/pulumi/commit/1bd1eaff50cacc027d8f478078548a0f00b3f1ae.

In the original engine, we had inputs -- specified by the program, and managed by the engine -- and outputs -- computed by the resource provider, and handed back to the engine on resource operations. To get a complete view of a resource's state, therefore, we had to merge inputs and outputs. That is, we didn't consider the provider responsible for inputs whatsoever. As a result, we ended up merging the two maps in lots of places, including when doing diffs, update operations, and so on.

Eventually, we realized much of this was wrong, due to our interaction with the TF bridge. (To be fair, we designed all of the above before the TF bridge was even in the picture, back when we wrote our own providers.) A few examples of troubles this caused included: 1) on the provider side, defaults are computed and injected into the state; 2) inputs may get rewritten arbitrarily inside the provider; and 3) certain inputs can't be strictly classified as inputs or outputs, because they can be either.

As a result, we shifted away from this in a few ways. First, diffing responsibility was moved into the provider, by way of the Diff RPC. Next, the inputs we feed into the actual CRUD operations changed from being the raw program inputs, to inputs computed by the provider inside of Check/Diff. Finally, we stopped thinking of inputs and outputs as being projected onto a single, unified bag, and simply embraced that they are different things. In fact, we have more than one view on inputs floating around inside of the engine, when you consider the inputs as produced by the program, versus inputs as specified by the provider in preparation for performing the CRUD operations.

In any case, I'm fairly certain any lingering calls to `All()` and/or `Merge()` are complete accidents of history, and represent codepaths we neglected to update as we performed the above changes. Something like https://github.com/pulumi/pulumi/pull/2659 is probably the solution, however, we will want to be very careful in evaluating whether the semantic changes implied here are correct.

And I agree we desperately need to document how all of this works, and do a careful audit to ensure it's actually working the way we think it should. It's been an organic evolution to get to where we are.

**Comment by lukehoban** [comment]
I tested the fixes in #2659, and it does indeed fix the issue in https://github.com/pulumi/pulumi-eks/issues/69, and doesn't fail any current tests in pulumi/pulumi.  I'm going to go ahead and merge that fix, as a have confidence it is "right".  However, there is still some chance this will uncover subtle dependencies on the previous behaviour somewhere upstack.  If we see any new issues as we bring this into nightly tests in other repos, we'll decide how best to proceed based on that data.
